hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/DiscoveryJoiner.java

```diff
@@ -71,6 +71,10 @@ public DiscoveryJoiner(Node node, DiscoveryService discoveryService, boolean use
 
     @Override
     protected Collection<Address> getPossibleAddressesForInitialJoin() {
+        if (maximumWaitingTimeBeforeJoinSeconds == 0) {
```

Comment:
Fix is confirmed, thanks for digging to the bottom of this issue @lprimak!

I think the issue was introduced in https://github.com/hazelcast/hazelcast/commit/ffc399a1c07cdb2ea128a0c5158ccc7a00abbe0a which changed `getPossibleAddressesForInitialJoin()`:
- prior to that commit, there was at least 1 call to `getPossibleAddresses` and its result would be returned
- after that commit, it is possible that we do not enter the deadline loop at all and return `Collections.emptyList()` which is wrong

Instead of treating `maximumWaitingTimeBeforeJoinSeconds == 0` as a special case, I 'd rather change this method's logic like this:

```
Collection<Address> possibleAddresses = null;
long deadLine = ... // code continues as is - possibleAddresses can be updated in the for loop
...

return (possibleAddresses == null) ? getPossibleAddresses() : possibleAddresses;
```

wdyt @lprimak ?

Reply:
sure, done